feat: Comment/description support, bug fixes and better test coverage

jimfulton opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [X] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-sqlalchemy/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [X] Ensure the tests and linter pass
- [X] Code coverage does not decrease (if any source code was changed)
- [X] Appropriate docs were updated (if necessary)

Fixes #89🦕

- Runs SQLAlchemy dialect-compliance tests (as system tests).
- 100% unit-test coverage.
- Support for table and column comments/descriptions (requiring SQLAlchemy 1.2 or higher).
- Fixes bugs found while debugging tests, including:
  - Handling of `in` queries.
  - String literals with special characters.
  - Use BIGNUMERIC when necessary.
  - Missing types: BIGINT, SMALLINT, Boolean, REAL, CHAR, NCHAR, VARCHAR, NVARCHAR, TEXT, VARBINARY, DECIMAL
  - Literal bytes, dates, times, datetimes, timestamps, and arrays.
  - Get view definitions.
- When executing parameterized queries, the new BigQuery DB API parameter syntax is used to pass type information.  This is helpful when the DB API can't determine type information from values, or can't determine it correctly.

Commit:
Fixed a dependency problem that caysed test failures in Python 3.6.

The source of the dependency bug is in old versions of google-cloud-core that
depend on too-old versions of google-api-core.

Commit:
Fixed a dependency problem that caysed test failures in Python 3.6.

The source of the dependency bug is in old versions of google-cloud-core that
depend on too-old versions of google-api-core.

Commit:
Started implementing SqlAlchemy dialect-compliance tests

Some tests are still failing, but
we're far enough along that we have the right shape, I think.

Commit:
Handle parameters in in

The moral equivalent of    "where foo in (@bar)", where bar is an array
which actually need to be  "where foo in unnest (@bar)".

Commit:
fixed regex to allow multi-character parameter names.

Doh!

Commit:
Fixed rendering of string literals.

Commit:
Provide default values for primary keys and other fixes...

Other fixes:
- Handle BIGINT
- Fix string leteral formatting (and start type-specific adaptations).

Commit:
Narrowed skips and license comment diagnosed reasons for skipped failures

Commit:
use Google license

Commit:
Fixed like (% and _) escapes.

The SQLAlchemy like convenience functions (e.g. ) escape incorrectly for
BigQuery, so re-escape.

Commit:
Handle BIGNUMERIC

Commit:
Skip tests that want to stpre floats as numeric.

We could make that work, if we want to. :)

Commit:
skip tests for offsets without limits. BigQuery doesn't allow that.

Commit:
BIGNUMERIC lets us handle many significant digits.test_many_significant_digits

Commit:
Skip LongNameBlowoutTest because it requires features (indexes, primary keys, etc., that BigQuery doesn't have.

Commit:
We have to rewrite these tests, because of forcing use_labels=True

https://github.com/googleapis/python-bigquery-sqlalchemy/issues/78

Commit:
BigQuery returns deleted and updated rows, which SQLAlchemy doesn't expect

Commit:
Provide a bigquery mock based on sqlite

So we don't have t mock at the api level.

Commit:
Don't force labels in select.

https://github.com/googleapis/python-bigquery-sqlalchemy/issues/78

Commit:
merge riversnake-fix-78

Commit:
No need to work around forced labels anymore.

Commit:
Updated sqlite-bases mock

- Run tests in temporary directory rather than sharing memory connections.
  Because simpler. :)
- Introduce cross-connection state and record queries in it, so tests can
  make assertions bout generated queries.

Commit:
Added test that types are injected into parameter placeholders.

Commit:
Oops, need to call super even when there are where clauses.

Commit:
When SQLAlchemy thinks it knows a parameter type, include it in the placeholder

The BigQuery Python Client supports an extended placeholder syntax
that includes type information, as in `%(foo:INT64)s` (named) or
`%(:INT64)s` (unnamed).

When we know the type, include it in the placeholder.

Commit:
Don't skip some numeric tests and skip some CTE tests

The numeric tests now tun since we started passing type info from sqla to bigquery.

The CTE tests test features that don't exist in bigquery.

Commit:
Can't coerce datetime to date.

Commit:
reenable some tests now that we pass type infor to bigquery

Commit:
Enable error on warnings -- a test depends on it.

Although the test isn't actually testing dialect code.  Maybe it
should be skipped instead.

Also set the profile test pasth to a more reasonable value, although
it doesn't seem to be used. <shrug>

Commit:
Added table and column comment support.

Also inlined colspecs code, because there wasn't much and it
facilitated separating literal processing into a function.

Commit:
get_schema_names should return all of the schema names.

The tests don't disagree. :)

Commit:
Retrieval of view definitions.

Commit:
handle dateish literals.

Commit:
BigQuery doesn't want schema names in column names.

Commit:
Allow mutating table descriptions/comments.

Commit:
skip tests that depend on BigQuery keeping track of column details.

Like precision, length, etc.

Commit:
Handle TIMESTAMP literals.

Commit:
Fixed some broken TIMESTAMP tests.

Also, removed some unneeded imports.

Commit:
enabled more tests.

Commit:
remove extra lines.

Commit:
blacken

Commit:
added missing test for api.py

Commit:
Ignore flake8 complaint about * import.

The names are known and used by pytest.

Commit:
lint and minor cleanup

Commit:
Removed unused variable.

Commit:
Get to 100% coverage in test_parse_url

- Test cases with one param set to reach an uncovered branch for write disposition.

- Test parsing false.

dandhlee committed:
chore(revert): revert preventing normalization (#132)

reverts previous commit for preventing normalization of versioning

Commit:
Don't need because we can't test temp tables

Because bigquery only supports temp tables in scripts.

Commit:
requirements.py -- It's used by compliance testing.

Commit:
Added Binary literal handling, and ...

also:

- Added missing small int.
- Renamed `process_literal` to `process_string_literal`
- some reformating.

Commit:
Updated fauxdb to correct for some sqlite differences.

Also, rearraged conftest and the faux_conn fixture a bit
to reuse sqlite connections and to go back to a memory connection.

Commit:
Fixed santy check

Commit:
blackened

Commit:
Added a test that exercises type handling for all the types.

(Replaced a simper test.)

Commit:
Added table reference tests

Commit:
test get_view_definition

Commit:
removed a private function that's never called.

Commit:
table and view list tests

Commit:
Moar unit tests and coverage

Commit:
use latest bg release, because we depend on changes there.

Commit:
removed breakpoint that shouldn't have been checked in.

Commit:
Bypass google authentication.

Also, simplify wireing of faux connections and clients.

Commit:
Test JSON deserialization.

Commit:
Test getting project id from authentication.

Commit:
removed some unneeded code.

String literals in bq can start with ".

Commit:
Simplify string and binary literal processors to not expect None.

It can't happen.

Commit:
Tested comment support

Commit:
Cleaned up comment handling.

BTW column.doc is a Python docs thing for the ORM, not a database level comment/description.

Commit:
generalized options handling.

Commit:
test dialect-options handling.

Commit:
constraints are ignored.

Commit:
Better binary-literal hack

Commit:
added tests for BIGNUMERIC.

Commit:
Major rework using pickle to overcome type differences.

Commit:
Added array support

Commit:
blacken

Commit:
replaced an unreachable branch with an assert.

Commit:
lint

Commit:
Enable generation of random primary keys when running compliance tests.

Because the tests expect tables to have primary keys that het provided somehow.

Commit:
Run the select a couple of ways and assert result.

Commit:
Only generate random primary keys when running compliance tests.

Commit:
addd fetchall, needed by a test.

Commit:
Port compliance tests

For unit-test coverage. :)

Commit:
record when array size is set.

Commit:
Added a tiny metadata convenience.

A convenience here, a convenience there ...

Commit:
Added arraysize tests.

Commit:
exercise labels

Commit:
Make sure label names are legal identifiers

Commit:
test disable quoting

Commit:
simplifed a test.

Commit:
Test forcing quote (even though quotes are forced anyway)

Commit:
Be more careful about doubled %s.

Commit:
ported test for %%s

Commit:
Added IN tests.

Commit:
handle UNNEST

Commit:
"inlined" some helper attrs

Commit:
moved setup_table to central location

Commit:
ported a test that deals with bind params with indeterminate types

Commit:
leverage setup_table

Commit:
added a test based on sqlalchemy.testing.suite.test_select.LikeFunctionsTest

Commit:
leverage setup_table

Commit:
Added test for labels in group by

Commit:
Added a test for compiling a single column.

Also,simplified other test using the metadata helper.

Commit:
Trying to compile a nameless column gives a meaningful error.

Commit:
Get 100% coverage of fauxdbi

Commit:
Simplified quote decision.

We always quote unless told not to.

Also, exclude a method for getting default primary-key values. It's only used in compliance tests.

Commit:
Specify proto 4 when pickling, to get predictable behavior across Python versions

Commit:
Some tests require sqlalchemy 1.3

Commit:
require 100% test coverage.

Commit:
need newer sqlalchemy and google-auth

sqlalchemy because comments.

google-auth because we need very recent bigquery.

Commit:
need sqlalchemy 1.2 because comments.

Commit:
blacken/lint

Commit:
Added code to cleanup schemas at start to avoid spurious errors.

Commit:
added copyright

Commit:
added copyright

Commit:
fixed copyright

Commit:
Updated the compliance setup

Commit:
added copyright

Commit:
added copyright

Commit:
added copyright

Commit:
added copyright

Commit:
added copyright

Commit:
blacken

Commit:
merge master

Commit:
Make operational errors more informative.

Commit:
Fixed better debugging support.

Commit:
lint first

Commit:
Don't use upsert to update comments.

The version of sqlite in the CI image doesn't support upsert.

Commit:
Older sqlite don't have true and false.

Commit:
Ignore test debugging info added to deal with sqlite3 differences.

Commit:
blacken

jimfulton commented:
These links may help:
https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_3_24/README.dialects.rst
https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_3_24/lib/sqlalchemy/engine/interfaces.py#L15

plamut reviewed (commented):
Based on what I've read - but with limited view of the big picture - the overall structure looks good. Having smaller commits made the review somewhat easier, but the PR is still _massive_. :slightly_smiling_face: 

Made a few comments at places where we might want to take another look, even if just for sanity.

jimfulton commented:
Sorry again about the size of the PR. If I'd used smaller PR's I'd have ended up with a fair bit of rework as I discovered more as I wrote more tests.  If I was adding features rather than mainly tests, I'd definitely have preferred smaller PRs.

Commit:
Added minimal requirement for google-api-core to be consistent with setup.py.

Commit:
Make the google-auth requirement consistent with the Python 3.6 test constraint.

The test constraint was really diven by the requitements of other
requirements, not this package. <shrug>

jimfulton commented on setup.py:
```diff
@@ -39,9 +54,9 @@ def readme():
+setuptools.setup(
     name=name,
-    version=version,
+    version=sic(version),
```
I didn't write this.  This came from: https://github.com/googleapis/python-bigquery-sqlalchemy/pull/125.


I just tried updating from master, which doesn't have this, but didn't get the revert. Weird.

I'll revert manually.

jimfulton replied:
But wait:

- #125 makes `sic` a noop for older versions of setuptools.
- But then it monkey-patches `packaging.version.Version`. Ugh.

jimfulton commented on setup.cfg:
```diff
@@ -20,8 +20,11 @@ universal = 1
 [db]
-default=bigquery://
-bigquery=bigquery://
+default=bigquery://precise-truck-742/test_pybigquery_sqla
```
It's not clear. This isn't very well documented.  I'll see if I can avoid including the project here.  

I'm a bit skeptical that an external developer will be ale to make the system and compliance tests work, but IDK. :)

I'll take a stab at documenting this issue.

jimfulton replied:
Done.

The default url no-longer has the project and I added a readme.

Commit:
Explain why we add where clauses in delete statements that lack them

but only during test cleanup.

jimfulton commented on tests/sqlalchemy_dialect_compliance/conftest.py:
```diff
@@ -0,0 +1,60 @@
+
+
+def visit_delete(self, delete_stmt, *args, **kw):
+    if delete_stmt._whereclause is None and "teardown" in set(
+        f.name for f in traceback.extract_stack()
+    ):
```
Comment:

```
# BigQuery requires delete statements to have where clauses. Other
# databases don't and sqlalchemy doesn't include where clauses when
# cleaning up test data.  So we add one when we see a delete without a
# where clause when tearing down tests.  We only do this during tear
# down, by inspecting the stack, because we don't want to hide bugs
# outside of test house-keeping.
```

added. :)

plamut replied:
Very good summary, thanks!

Commit:
Added a missing _ and explained \d+s in a regex

jimfulton commented on pybigquery/sqlalchemy_bigquery.py:
```diff
@@ -156,19 +145,68 @@ def create_cursor(self):
+    def pre_exec(
+        self,
+        in_sub=re.compile(
+            r" IN UNNEST\(\[ "
+            r"(%\([^)]+\d+\)s(, %\([^)]+_\d+\)s)+)?"  # Placeholders
+            ":([A-Z0-9]+)"  # Type
+            r" \]\)"
```
> Also, what digits are we trying to match?
> 
> ```
> (%\([^)]+\d+\)s
> ```
> 
> Are these parameter indexes SQLAlchemy puts in these placeholders?
> 
> I also noticed that in the first part, that "numeric" bit is not prefixed by an underscore, while in the second (optional) part it is: `_\d+` - on purpose or a typo?

Yes, you're write, on what they're bout and the typo.  I added the missing `_` as well as this comment.

```
        # In the placeholder part of the regex above, the `_\d+
        # suffixes refect that when an array parameter is expanded,
        # numeric suffixes are added.  For example, a placeholder like
        # `%(foo)s` gets expaneded to `%(foo_0)s, `%(foo_1)s, ...`.
```

jimfulton replied:
> > (nit)
> > I find it easier to read such strings if written as multiline string literals, because no opening and closing quotes in the pattern to distract the eyes. :) And in `re.VERBOSE` mode whitespace is already ignored, allowing for useful comments. The downside is that that those few space characters will need to be explicitly escaped, though.
> > Feel free to ignore if your brain is wired differently, or if you don't see enough value this change would bring.
> 
> I don't fully follow your comment.
> 
> How would you have written this?

I tried this.
- I found the space escaping distracting.
- It made the commends harder to see. This may be a personal (editor) problem.
- I trues 3 times and wasn't able to not break the regex. I must have been escaping the spaces wrong.

<shrug>

plamut replied:
Fair enough, let's leave it as is then. I personally found the verbose mode useful when writing some very complex replacement rules in `synth.py`, but having to explicitly escape any expected whitespace can indeed be a significant drawback.

Commit:
Fixed: didn't properly handle expansion of single-item arrays.

Commit:
Added test for expansion of single-element arrays

Commit:
Avoid having to use an `_` to ignore a re group we don't care about.

jimfulton commented on pybigquery/sqlalchemy_bigquery.py:
```diff
@@ -156,19 +145,68 @@ def create_cursor(self):
+        # purposes.)
+
+        def repl(m):
+            placeholders, _, type_ = m.groups()
```
You found a bug!

In making that pattern non-matching, I realized I wasn't handling expanding single-element arrays correctly. Test and fix added.

Commit:
Added some comments for the tricky _get_field method.

jimfulton commented on tests/unit/fauxdbi.py:
```diff
@@ -54,45 +272,150 @@ def executemany(self, operation, parameters_list):
+            custom = columns.get(name)
+            if custom:
+                return self._get_field(
+                    **dict(name=name, type=type, notnull=notnull, **custom)
```
We express the kwargs this way to allow custom values to overrise
name, type and nonnull, if necessary.

I added some comments.

jimfulton replied:
Oops. I had in my head: `dict(dict(name='x'), **dict(name='y'))`

Since we're not relying on overriding (I probably was or planned to at some point), I just got rid of it.

Thanks!

Commit:
Commented the pickle-protocol 4 prefixes we're looking for.

Commit:
explain use of pickle.

jimfulton commented on testing/constraints-3.6.txt:
```diff
@@ -4,6 +4,6 @@
-sqlalchemy==1.1.9
-google-auth==1.14.0
-google-cloud-bigquery==1.12.0
+sqlalchemy==1.2.0
+google-auth==1.24.0
+google-cloud-bigquery==2.15.0
```
Done.

jimfulton commented on tests/unit/fauxdbi.py:
```diff
@@ -54,45 +272,150 @@ def executemany(self, operation, parameters_list):
-        self.cursor.fetchmany(size or self.arraysize)
+        return [
+            (
+                pickle.loads(v.encode("latin1"))
+                if isinstance(v, str) and v[:2] == "\x80\x04" and v[-1] == "."
+                else pickle.loads(base64.b16decode(v))
+                if isinstance(v, str) and v[:4] == "8004" and v[-2:] == "2E"
+                else v
```
Yup. Added.

These are pickle protocol 4 prefixes.

I added comments, including a couple of paragraphs near the top of the class explaining how and why we're using pickle.

Commit:
typo

Commit:
BigQuery 2.15 wants google-api-core >= 1.23.0

Commit:
blacken

Commit:
don't need to call out the project in the test config.

We get it from the creds.

Commit:
Try sleeping between tests to see if we can avoid some weird failures.

Commit:
some minimal docs on the dialect tests.

plamut commented on tests/sqlalchemy_dialect_compliance/README.rst:
```diff
@@ -0,0 +1,12 @@
+==================================
+
+SQLAlchemy provides reusable tests that test that SQLAlchemy dialects
+work prooerly.  This directory applies these tests to the BigQuery
```
```suggestion
work properly. This directory applies these tests to the BigQuery
```

plamut commented on noxfile.py:
```diff
@@ -169,6 +170,54 @@ def system(session):
+    system_test_folder_path = os.path.join("tests", "sqlalchemy_dialect_compliance")
+
+    # Check the value of `RUN_SYSTEM_TESTS` env var. It defaults to true.
+    if os.environ.get("RUN_COMPLIANCE_TESTS", "true") == "false":
```
I think that's generated in other libraries' noxfiles, thus we keep it here, too, even though we have a manual `noxile.py`.
(in `google-cloud-bigquery` at least, but presumably here, too). Let's leave it as is then and not worry too much. :)

plamut reviewed (changes requested):
Good that we caught and fixed a bug!

The changes look good, I just found two more typos and what seems like a bug with overriding default kwarg values in `fauxdbi`. No tests currently seem to use that, thus no failure.

Commit:
Update tests/sqlalchemy_dialect_compliance/README.rst

Co-authored-by: Peter Lamut <plamut@users.noreply.github.com>

Commit:
Update tests/sqlalchemy_dialect_compliance/README.rst

Co-authored-by: Peter Lamut <plamut@users.noreply.github.com>

Commit:
simplify by getting rid of unnecessary and broken overrriding.

Commit:
merge upstream

plamut commented on tests/unit/fauxdbi.py:
```diff
@@ -350,12 +350,7 @@ def _get_field(
-                    # necessary.
-                    **dict(name=name, type=type, notnull=notnull, **custom)
-                )
+                return self._get_field(name=name, type=type, notnull=notnull, **custom)
```
If I understood correctly, `custom`, by design, can never contain `name`, `type`, or `notnull` keys?

plamut replied:
You probably know better. :)

If we want to use overriding, the code needs to be changed to avoid errors (duplicate keyword arguments), otherwise we can leave it as is.

jimfulton replied:
I removed the overriding.  If we need it later, we can add it then.

Commit:
Try sleeping longer to avoid races between tests.

shollyman commented on tests/sqlalchemy_dialect_compliance/conftest.py:
```diff
@@ -0,0 +1,74 @@
+    for schema in "test_schema", "test_pybigquery_sqla":
+        for table_item in client.list_tables(f"{client.project}.{schema}"):
+            table_id = table_item.table_id
+            client.query(
```
Your timing issues may be related to this.  It seems that you're invoking the query, but not waiting for it to complete.  Possibly just racy?

Commit:
make sure we wait for the drops at the beginning of the session.

And get rid of the sleep, at least termporarily.

Commit:
rerun on table not found

Commit:
Try simpler regex

Commit:
Try different error

Commit:
don't sleep between tests.

It didn't buy us anything.

Commit:
blacken

Commit:
removed unused import

jimfulton commented:
@plamut are you requesting changes? :)

plamut reviewed (approved):
No, it was just late yesterday and didn't get to submit an approval - doing this now. :)
Looks like a good piece of work, thanks!

---
Disclaimer - I don't know all the ins and outs of writing a new SQAlchemy dialect, as I have not studied all the requirements (especially not some less- or non-documented ones) . Did the reviews on a best-effort basis.

Any additional pair of eyes would be welcome, if anybody else has some free capacity.
